_drafts/zio-fibers.md

```diff
@@ -0,0 +1,290 @@
+date: 2021-06-31 
+header:
+  image: "/images/blog cover.jpg"
+tags: []
+excerpt: "Many libraries implement the effect pattern in the Scala ecosystem, and every one has its own concurrency model. First, let's introduce the ZIO library and its implementation of the fiber model."
+---
+
+Many libraries implement the effect pattern in the Scala ecosystem: Cats Effect, Monix, and ZIO, just to list some. Every of these implements its own concurrency model. For example. Cats Effect and ZIO both rely on _fibers_. In the articles [Cats Effect 3 - Introduction to Fibers](https://blog.rockthejvm.com/cats-effect-fibers/) and [Cats Effect 3 - Racing IOs](https://blog.rockthejvm.com/cats-effect-racing-fibers/), we introduced the fiber model adopted by the Cats Effect library. Now, it's time to analyze the ZIO library and its implementation of the fiber model.
+
+## 1. Background and Setup
+
+We live in a beautiful world where Scala 3 is the actual major release of our loved programming language. So, we will use Scala 3 throughout the article. Moreover, we will need the dependency from the ZIO library:
+
+```sbt
+libraryDependencies += "dev.zio" %% "zio" % "1.0.9"
+```
+
+Nothing else will be required.
+
+## 2. The Effect Pattern
+
+Before talking about ZIO fibers, we need to have the notion of the _Effect Pattern_. The main objective of any effect library is to deal with statements that can lead to side effects. The _substitution model_ is the fundamental building block of functional programming. Unfortunately, **we cannot model side effects using the substitution model because we cannot substitute functions with  their results**, varying from information other than simple functions' inputs. Such functions are often called _impure_:
+
+```scala
+// The type of println is String => Unit. The program prints the given String to the console
+val result = println("The meaning of life is 42")
+// Using the substitution model, we try to substitute the result of the println execution to the variable
+val result: Unit = ()
+//...however, after the substitution, the meaning of the program completely changed
+```
+
+So, it comes the Effect Pattern. The pattern aims to model side effects with the concept of effect. **An effect is a blueprint of statements that can produce a side effect, not the result itself**. So, when we instantiate an effect with some statements, we don't execute anything: We are just describing what the code inside the effect will perform once executed:
+
+```scala
+// This code doesn't print anything to the console. It's just a blueprint
+val result = ZIO.succeed(println("The meaning of life is 42"))
+```
+
+
+The above code doesn't print anything to the console; it just describes what we want to achieve with the code, not how we execute it.
+
+Moreover, the Effect Pattern adds two more conditions to the whole story:
+
+1. The type of the effect must describe the type returned by the effect once executed and what kind of side effect it represents.
```

Comment:
This sentence uses "effect" with 3 different meanings. I'd recommend something like

"The type of the effect must describe a) the _kind of computation_ it executes, b) the _value_ it contains or produces (usually via a generic type argument)."